crypto/fipsmodule/curve25519/curve25519.c

```diff
@@ -169,35 +328,143 @@ int ED25519_sign_no_self_test(uint8_t out_sig[ED25519_SIGNATURE_LEN],
+    const uint8_t public_key[ED25519_PUBLIC_KEY_LEN]) {
+  return ed25519_verify_internal(ED25519_ALG, message, message_len, signature,
+                                 public_key, NULL, 0);
+}
+
+int ED25519ctx_verify(const uint8_t *message, size_t message_len,
+                      const uint8_t signature[ED25519_SIGNATURE_LEN],
+                      const uint8_t public_key[ED25519_PUBLIC_KEY_LEN],
+                      const uint8_t *context, size_t context_len) {
+  FIPS_service_indicator_lock_state();
   boringssl_ensure_eddsa_self_test();
-  return ED25519_verify_no_self_test(message, message_len, signature, public_key);
+  int res = ED25519ctx_verify_no_self_test(message, message_len, signature,
+                                           public_key, context, context_len);
+  FIPS_service_indicator_unlock_state();
+  return res;
+}
+
+int ED25519ctx_verify_no_self_test(
+    const uint8_t *message, size_t message_len,
+    const uint8_t signature[ED25519_SIGNATURE_LEN],
+    const uint8_t public_key[ED25519_PUBLIC_KEY_LEN], const uint8_t *context,
+    size_t context_len) {
+  return ed25519_verify_internal(ED25519CTX_ALG, message, message_len,
+                                 signature, public_key, context, context_len);
+}
+
+int ED25519ph_verify(const uint8_t *message, size_t message_len,
+                     const uint8_t signature[ED25519_SIGNATURE_LEN],
+                     const uint8_t public_key[ED25519_PUBLIC_KEY_LEN],
+                     const uint8_t *context, size_t context_len) {
+  FIPS_service_indicator_lock_state();
+  boringssl_ensure_hasheddsa_self_test();
+  int res = ED25519ph_verify_no_self_test(message, message_len, signature,
+                                          public_key, context, context_len);
+  FIPS_service_indicator_unlock_state();
+  if (res) {
+    FIPS_service_indicator_update_state();
+  }
+  return res;
+}
+
+int ED25519ph_verify_no_self_test(
+    const uint8_t *message, size_t message_len,
+    const uint8_t signature[ED25519_SIGNATURE_LEN],
+    const uint8_t public_key[ED25519_PUBLIC_KEY_LEN], const uint8_t *context,
+    size_t context_len) {
+  uint8_t digest[SHA512_DIGEST_LENGTH] = {0};
+  SHA512_CTX ctx;
+  SHA512_Init(&ctx);
+  SHA512_Update(&ctx, message, message_len);
+  SHA512_Final(&digest[0], &ctx);
+  return ED25519ph_verify_digest_no_self_test(digest, signature, public_key,
+                                              context, context_len);
+}
+
+int ED25519ph_verify_digest(const uint8_t digest[SHA512_DIGEST_LENGTH],
+                            const uint8_t signature[ED25519_SIGNATURE_LEN],
+                            const uint8_t public_key[ED25519_PUBLIC_KEY_LEN],
+                            const uint8_t *context, size_t context_len) {
+  FIPS_service_indicator_lock_state();
+  boringssl_ensure_hasheddsa_self_test();
+  int res = ED25519ph_verify_digest_no_self_test(
+      &digest[0], signature, public_key, context, context_len);
```

Comment:
Curious why you do `&digest[0]` instead of just giving it the pointer?

Reply:
Just out of habit from how you would get the sizeof the element itself normally. I can update.